x/valset-pref/validator_set.go

```diff
@@ -170,49 +214,55 @@ func (k Keeper) PreformRedelegation(ctx sdk.Context, delegator sdk.AccAddress, e
 	}
 
 	// calculate the difference between two sets
-	var diffValSet []*valSet
+	var diffValSets []*valSet
+	// Matt: Does this mean that the diffAmount can be negative?
 	for i, newVals := range existingValSet {
 		diffAmount := newVals.amount.Sub(newValSet[i].amount)
 
 		diff_val := valSet{
 			valAddr: newVals.valAddr,
 			amount:  diffAmount,
 		}
-		diffValSet = append(diffValSet, &diff_val)
+		diffValSets = append(diffValSets, &diff_val)
 	}
 
 	// Algorithm starts here
-	for _, diff_val := range diffValSet {
-		for diff_val.amount.TruncateDec().GT(sdk.NewDec(0)) {
-			source_val := diff_val.valAddr
-			target_val, idx := k.FindMin(diffValSet, source_val)
+	// Matt: we need more detailed documnetaiton on the algorithm here
+	// Matt: We're running in computation of O(n^3) here, two for loops here and then another iteration in the `FindMinAmtValSetExcept`
+	// What's the reason we'we need to find min for every iteration rather than simply using whats in the array in order?
```

Comment:
It's because after every redelegation there is going to be a new min. for instance 

diff_arr = [ValA: 10, ValB: 6, ValC: 4, ValD: -4, ValE: -4, ValF: -12]

the algorithm runs as follow 
```
loop through diff_arr (only the positive numbers): 
**** 1st iteration*****
    sourceVal = get the first validator Amount (valA in our case) 
    targetVal = Find the minimum amount in the entire diff_arr (valF in our case) 
    redelegate from valA --> valF (10) 
    update valA = 0, valF = -2 

***** 2nd iteration **** 
   sourceVal = get the 2nd validator Value (valB) 
   targetVal = Find the minimum amount in the entire diff_arr (ValD or ValE in our case) lets say we pickD 
   redelegate from valB --> valD (4) 
   update valB = 2, valD = 0 

..... We continue doing until all the validator values [valA..valF is 0, or if the minimum value we find is 0 because if minAmount is 0 there is no more redelegation possible) 

```

Reply:
Did we ever get back to the necessity to use the min with the trade off of doing computation of O(n^3) vs just using the orders in array to make the computation O(n^2)?

Reply:
yes, i forgot to mention this but i did try without min, however i still have to iterate through the diffSet to find the first negative number. 

I can try to split up diffSet positive and negative numbers but that'll still require looping through both loops which can get complex.

Reply:
I still think this complexity is solvable by creating two arrays : negative array and positive array here https://github.com/osmosis-labs/osmosis/pull/4143/files#diff-4422f5e761e99a0e29d2ba7017ece4aa6b486b5ee24b18fb58fcbc9c26a8e7a6R239-R247 and then using those two arrays instead of having to iterate over them again

Reply:
Issue created for this so that we dont delay v15 https://github.com/osmosis-labs/osmosis/issues/4256